### DISCUSSION TITLE
Question: Properties and parameters validation

### DISCUSSION THREAD

**Discussion by GreatTeacherBasshead** [psake/psake#379] [Q&A]
## Context

1. I want to use a parameter passed via command line to initialize some properties which are used in multiple tasks, instead of initializing them in each task.
`Invoke-psake script.ps1 -parameters @{"baseDir" = "...some_path_here..."}`

2. And I want to validate that parameter first.

## Example
```
Properties {
$reusedPath = Join-Path $baseDir "smth"
}

Task Default -Depends Validate, TaskDB, TaskServer

Task Validate {
Assert (![string]::IsNullOrEmpty($baseDir)) '$baseDir was null or empty'
Assert (Test-Path $baseDir) "`$baseDir does not exist"
}

Task TaskDB {
$x = Join-Path $reusedPath "smth1"
}

Task TaskServer {
$x = Join-Path $reusedPath "smth2"
}
```

## Problem
Properties are initialized before any task. So my validation task doesn't do any work at all. For instance, if I don't pass **$baseDir** as a parameter or leave it empty, I will not reach Validate task. It will fail on a step of Properties initialization (`$reusedPath = Join-Path $baseDir "smth"`) with really not informative error:
`~~~~~~~~~~~ [<<==>>] Exception: Cannot bind argument to parameter 'Path' because it is null.`

## Question
Is there way to validate parameters before Properties initialization? Or is there other way how to achieve the same goal?

## Environment
* Module version used: 4.7.0
* Operating System and PowerShell version: Windows 10, PowerShel 5.1

**Comment by HeyItsGilbert** [comment]
Since this is a how-to question (and tagged as one), I'm moving it to Discussions where Q&A lives now. Short version: a Properties block is evaluated before any task runs, so the Join-Path against $baseDir in there happens before your Validate task ever gets a chance to assert on it. If you want validation to gate the property setup, either do the assert inside the Properties block itself, or compute the derived properties inside a task that depends on Validate rather than in the global Properties block. Happy to go deeper in the discussion thread.
